### PR TITLE
Update @wordpress and React packages [MAILPOET-5102]

### PR DESCRIPTION
**Pull request by costasovo**
## Description
This PR updates @wordpres/* and React packages. React update is included because some of the wordpress packages [have already updated to React 18](https://github.com/WordPress/gutenberg/pull/45235) and use the new functionality.

Note: After the update, there is the dev console for dev builds.
```
Warning: ReactDOM.render is no longer supported in React 18. Use createRoot instead. Until you switch to the new API, your app will behave as if it's running React 17. Learn more: https://reactjs.org/link/switch-to-createroot
```
This is expected for now and will be addressed in a follow-up ticket. Switching to the new render API will turn on the automatic batching feature, and it may cause some issues.

## Code review notes

_N/A_

## QA notes

This changes a lot of JS libraries we build on, so potentially, this could cause issues in the whole plugin.
Five parts of the plugin are influenced the most because they heavily use these packages:

1. Automation - some visual changes are expected
2. Form Editor - some visual changes are expected
3. Plugin settings
4. Dynamic segments editor (the form for creating segments)
5. Settings panel of the coupon block - some small visual changes are expected

## Linked PRs

https://github.com/mailpoet/mailpoet-premium/pull/692 - related changes in the premium plugin
https://github.com/mailpoet/wordpress-images/pull/43 - update of the image for CI builds

## Linked tickets

[MAILPOET-5102]

## After-merge notes

This also updates the node version in the dev environment so ping devs to update the local node and rebuild the WordPress image in their dev env.

[MAILPOET-5102]: https://mailpoet.atlassian.net/browse/MAILPOET-5102?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: mailpoet/assets/js/src/automation/components/option-button/index.tsx [hunk 1]
```diff
@@ -38,7 +38,7 @@ export function OptionButton({
       </Button>
       {slots.length > 0 &&
         slots.map(({ key, slot }) => (
-          <Fragment key={`slot-${key}`}>{slot}</Fragment>
+          <Fragment key={`slot-${key}`}>{slot()}</Fragment>
```
Comment (JanJakes) [inline mailpoet/assets/js/src/automation/components/option-button/index.tsx]: Oh, did this work before? I also wonder if there's a difference between `{slot()}` and `<Slot>`...

File: mailpoet/assets/js/src/automation/editor/components/header/document_actions.tsx [hunk 2]
```diff
@@ -48,9 +40,9 @@ function DocumentActions({ children }): JSX.Element {
         {children && (
           <Dropdown
             popoverProps={{
-              anchorRef: titleRef.current,
+              placement: 'bottom-start',
+              anchor: titleRef.current,
             }}
-            position="bottom center"
```
Comment (JanJakes) [inline mailpoet/assets/js/src/automation/editor/components/header/document_actions.tsx:53]: More of a note to self to check if the dropdown positioning in the UI somehow changed.

File: mailpoet/.babelrc [hunk 3]
```diff
@@ -1,8 +1,8 @@
 {
   "presets": [
-    "@babel/preset-typescript",
     ["@babel/preset-react", { "runtime": "automatic" }],
-    "@babel/preset-env"
+    "@babel/preset-env",
+    ["@babel/preset-typescript", { "allowDeclareFields": true }]
```
Comment (JanJakes) [inline mailpoet/.babelrc:5]: I wonder whether we should review the preset order in any other config (for any reason)...

File: mailpoet/assets/js/src/form_editor/components/sidebar/sidebar.tsx [hunk 4]
```diff
@@ -1,9 +1,21 @@
...
 import { useDispatch, useSelect } from '@wordpress/data';
+import {
+  ReduxStoreConfig,
+  StoreDescriptor,
+} from '@wordpress/data/build-types/types';
 import { DefaultSidebar } from './default_sidebar';
 import { PlacementSettingsSidebar } from './placement_settings_sidebar';
 import { store } from '../../store';
 
+// workaround for block editor store useSelect
```
Comment (JanJakes) [inline mailpoet/assets/js/src/form_editor/components/sidebar/sidebar.tsx]: Was this later removed (moved to `types/...` as per https://github.com/mailpoet/mailpoet/pull/4786/commits/5638e314cd633498efe6a247dcc6328775d77175)? Need to check.

File: mailpoet/assets/js/src/form_editor/components/sidebar/default_sidebar.tsx [hunk 5]
```diff
@@ -8,6 +12,14 @@ import { BlockSettings } from './block_settings.jsx';
 import { SidebarHeader } from './sidebar_header';
 import { store } from '../../store';
 
+// workaround for block editor store useSelect
```
Comment (JanJakes) [inline mailpoet/assets/js/src/form_editor/components/sidebar/default_sidebar.tsx]: Was this later removed (moved to `types/...` as per https://github.com/mailpoet/mailpoet/pull/4786/commits/5638e314cd633498efe6a247dcc6328775d77175)? Need to check.

**Review by JanJakes — COMMENTED** [review]
Great job! 👏

I added a few small notes. I can verify these myself, given that you'll be AFK, @costasovo.
Thanks!

**Review by JanJakes — DISMISSED** [review]

**Comment by Aschepikov** [comment]
Console error when I am pressing on add a condition

image

and some tiny issues 

1. In the automations incorrect tags list
image
On prod 
image

2. Automations background is not blurred when an element is deleting
image
On prod
image

3. On the prod all lists in the trigger, add/remove tags and lists are open when you are adding those actions, on this branch they are closed, I am not sure how it is better
image
On prod
image

4. Text in some field names is in capslock
image
image
On prod
image

5. Name of the automation is blue on this branch
image
On prod
image

6. Issue with the + button in the form editor
image
On prod
image

7. No space in the size heading
image
On prod
image

8. Font family is bold on prod
image
On prod
image

9. Lists in the colors and message colors are different from the prod, and should be disabled when there is no chosen color
image

10. Different UI of buttons in the color settings
image
image
On prod
image

11. Different UI in the template settings
image

**Comment by JanJakes** [comment]
Thanks, @Aschepikov! Great catches 👏

> Console error when I am pressing on add a condition

This is strange, I can't replicate it. Tried both with the premium plugin enabled and disabled. Did you have the correct premium plugin JS (caches and stuff) when you were testing that? I don't know how to reproduce it. On your screenshot, I don't see the "all of", "any of" switch, so I'd guess the premium plugin is not active, or the JS is not working. Should I check it on your website ❓

Issues:

> 1. In the automations incorrect tags list

Trying to play with the tags, but can't reproduce this. Do you have any steps to reproduce? Or should I check it on your website❓

> 2. Automations background is not blurred when an element is deleting

Indeed, it seems that [Gutenberg doesn't create the blur anymore](https://wordpress.github.io/gutenberg/?path=/story/components-modal--default). That's fine, we'll follow whatever they do ✅

> 3. On the prod all lists in the trigger, add/remove tags and lists are open when you are adding those actions, on this branch they are closed, I am not sure how it is better

That was a bug in Gutenberg components, now it's fixed, which is desired (enabled left image, disabled right image) ✅

> 4. Text in some field names is in capslock

That seems to be the Gutenberg style for labels in the sidebar. We can see it in the post editor itself, as well as in the individual components, [such as InputControl](https://wordpress.github.io/gutenberg/?path=/story/components-experimental-inputcontrol--with-help-text) ✅

> 5. Name of the automation is blue on this branch

Good catch, this seems to differ from Gutenberg. I need to investigate 🐞

> 6. Issue with the + button in the form editor

That looks like a bug 🐞

> 7. No space in the size heading 

Also a bug 🐞

> 8. Font family is bold on prod

That seems to be a style of [the underlying Gutenberg component](https://wordpress.github.io/gutenberg/?path=/story/components-customselectcontrol--default) ✅

> 9. Lists in the colors and message colors are different from the prod, and should be disabled when there is no chosen color

This seems to work exactly as in the WP post editor for me, so I'd say it's OK ✅

<div style="display: grid; gap: 32px;">
Screenshot 2023-03-15 at 15 15 19

Screenshot 2023-03-15 at 15 15 28
</div>

> 10. Different UI of buttons in the color settings

Same here, looks like WP post editor, so I think OK ✅

> 11. Different UI in the template settings

We could fix this, but isn't the new UI actually better? It's probably more translation-friendly as well (more space). I'd keep it on two lines ✅

**Comment by Aschepikov** [comment]
@JanJakes  For the first error you can check it here https://alex.qawp.net/wp-admin/admin.php?page=mailpoet-segments#/new-segment
I reinstalled plugins and cleaned all cache, I mentioned in slack that it is reproducible just on this site, maybe some problem with site

**Comment by Aschepikov** [comment]
With tags, I can't reproduce it after reinstalling too, maybe there was a lag with installing or a cache problem. 
I agree with all your comments. 

Just the 4 with capslock, there are some headings without capslock and I remember that there was a ticket with fixing some capslock headings in forms editors. Looks a lit bit strange for me this capslock, but yes I checked it in wp posts and they have it too.
image
image

**Comment by JanJakes** [comment]
@Aschepikov I fixed 5, 6, 7, and also 11, and another small issue.

> Just the 4 with capslock, there are some headings without capslock and I remember that there was a ticket with fixing some capslock headings in forms editors. Looks a lit bit strange for me this capslock, but yes I checked it in wp posts and they have it too.

Headers in the panels are not uppercased. It's only labels that are uppercased.

Just need to check the first issue and all else should be fine.

**Comment by JanJakes** [comment]
@Aschepikov I also likely fixed number 1 — improved the code to avoid any race conditions or duplicates. I also found a new issue (that the first time the "add tag" select was focused, it showed no autocomplete) and fixed it too.

**Review by Aschepikov — APPROVED** [review]
Great job!

**Pull request merged**